Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm

```diff
@@ -28,25 +28,100 @@
 #endif
 
 #import "config.h"
+#import "MessageSenderInlines.h"
 #import "WebExtensionAPINamespace.h"
 #import "WebExtensionAPIWebNavigation.h"
+#import "WebExtensionContext.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionFrameIdentifier.h"
+#import "WebExtensionFrameParameters.h"
+#import "WebExtensionUtilities.h"
+#import "WebProcess.h"
 
 #if ENABLE(WK_WEB_EXTENSIONS)
 
 namespace WebKit {
 
+static NSString *tabIdKey = @"tabId";
+static NSString *frameIdKey = @"frameId";
+static NSString *parentFrameIdKey = @"parentFrameId";
+static NSString *errorOccurredKey = @"errorOccurred";
+static NSString *urlKey = @"url";
+
+static NSDictionary *toWebAPI(const WebExtensionFrameParameters& frameInfo)
+{
+    NSMutableDictionary *result = [NSMutableDictionary dictionary];
+
+    result[errorOccurredKey] = frameInfo.errorOccurred ? @(frameInfo.errorOccurred.value()) : @NO;
+    result[parentFrameIdKey] = @(toWebAPI(frameInfo.parentFrameIdentifier));
+    result[urlKey] = frameInfo.url ? (NSString *)frameInfo.url.value().string() : @"";
+
+    if (frameInfo.frameIdentifier)
+        result[frameIdKey] = @(toWebAPI(frameInfo.frameIdentifier.value()));
+
+    return [result copy];
+}
+
+static NSArray<NSDictionary *> *toWebAPI(const Vector<WebExtensionFrameParameters>& allFrames)
+{
+    NSMutableArray *result = [NSMutableArray arrayWithCapacity:allFrames.size()];
+
+    for (auto frame : allFrames)
+        [result addObject:toWebAPI(frame)];
+
+    return [result copy];
```

Comment:
```
    return createNSArray(allFrames, [](auto& frame) {
        return toWebAPI(frame);
    }).get();
```

Reply:
Might need `#import <wtf/cocoa/VectorCocoa.h>` for this.

Reply:
Neat!